From: Josh Law <objecting@objecting.org>
Subject: lib/bootconfig: fix snprintf truncation check in xbc_node_compose_key_after()
Date: Thu, 12 Mar 2026 19:11:43 +0000

snprintf() returns the number of characters that would have been written
excluding the NUL terminator.  Output is truncated when the return value
is >= the buffer size, not just > the buffer size.

When ret == size, the current code takes the non-truncated path, advancing
buf by ret and reducing size to 0.  This is wrong because the output was
actually truncated (the last character was replaced by NUL).  Fix by using
>= so the truncation path is taken correctly.

Link: https://lkml.kernel.org/r/20260312191143.28719-4-objecting@objecting.org
Signed-off-by: Josh Law <objecting@objecting.org>
Reviewed-by: Andrew Morton <akpm@linux-foundation.org>
Cc: Masami Hiramatsu <mhiramat@kernel.org>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 lib/bootconfig.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/lib/bootconfig.c~lib-bootconfig-fix-snprintf-truncation-check-in-xbc_node_compose_key_after
+++ a/lib/bootconfig.c
@@ -316,7 +316,7 @@ int __init xbc_node_compose_key_after(st
 			       depth ? "." : "");
 		if (ret < 0)
 			return ret;
-		if (ret > size) {
+		if (ret >= size) {
 			size = 0;
 		} else {
 			size -= ret;
_
